src/common/transformations/src/transformations/common_optimizations/pull_through_reduce.cpp

```diff
@@ -0,0 +1,245 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/common_optimizations/pull_through_reduce.hpp"
+
+#include <memory>
+#include <ngraph/pattern/op/or.hpp>
+#include <openvino/core/validation_util.hpp>
+#include <openvino/op/util/reduction_base.hpp>
+#include <openvino/opsets/opset9.hpp>
+#include <vector>
+
+#include "itt.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "transformations/utils/utils.hpp"
+
+using namespace ov;
+using namespace std;
+
+namespace {
+// Adjust axes of Unsqueeze/Reduce ops after Unsqueeze pulling
+// For example if we have:
+//          input(shape={5,10,15})
+//                 |
+//          Unsqueeze(axes=[0,3]) -> output_shape = {1,5,10,1,15}
+//                 |
+// ReduceOp(axes=[2,4], keep_dims=false) -> output_shape = {1,5,1}
+// after pulling it will be:
+//          input(shape={5,10,15})
+//                 |
+// ReduceOp(axes=[1,2], keep_dims=false) -> output_shape = {5}
+//                 |
+//          Unsqueeze(axes=[0,2]) -> output_shape = {1,5,1}
+const std::vector<int64_t> adjust_axes(const std::vector<int64_t>& axes_to_align,
+                                       const std::vector<int64_t>& offset_axes) {
+    auto number_of_axes_less_than = [&offset_axes](const int64_t current_axis) {
+        return std::count_if(std::begin(offset_axes),
+                             std::end(offset_axes),
+                             [&current_axis](const int64_t excluded_axis) {
+                                 return excluded_axis < current_axis;
+                             });
+    };
+    std::vector<int64_t> result;
+    for (const auto& axis : axes_to_align) {
+        result.push_back(axis - number_of_axes_less_than(axis));
+    }
+    return result;
+}
+
+// Try to represent given Reshape node via Unsqueeze and calculate axes of such Unsqueeze
+// For example if we have Reshape(input_shape={5,10,15}, target_shape={1,5,10,15,1}),
+// it can be represented via Unsqueeze with axes=[0,3]
+std::vector<int64_t> try_get_unsqueeze_axes_from_reshape(const Shape& reshape_shape, const Shape& reduce_shape) {
+    if (reshape_shape.size() < reduce_shape.size()) {
+        return {};
+    }
+    const auto it = std::search(std::begin(reshape_shape),
+                                std::end(reshape_shape),
+                                std::begin(reduce_shape),
+                                std::end(reduce_shape));
+    if (it == std::end(reshape_shape)) {
+        return {};
+    }
+    const auto begin_pos = it - std::begin(reshape_shape);
+    const auto end_pos = begin_pos + reduce_shape.size();
+    auto equal_one = [](size_t elem) {
+        return elem == 1;
+    };
+    if (std::all_of(std::begin(reshape_shape), std::begin(reshape_shape) + begin_pos, equal_one) &&
+        std::all_of(begin(reshape_shape) + end_pos, std::end(reshape_shape), equal_one)) {
```

Comment:
I'm sorry if I don't get smth, but does it only allow for reshape shapes like 1, 1, ..., *reduce_shape, 1, 1, 1?

Reply:
provided more detailed decription and the tikcet